### ISSUE TITLE
proxy failed to connect to istiod.istio-system.svc:15012 with user-specified secret

### ISSUE THREAD

**Issue by patrickppeng** [istio/istio#28939]
**Bug description**
I’m trying to upgrade from Istio 1.5.4 to 1.7.4 and looking for help on how to use user-specified certificate for in-cluster mtls communication. We are using our own enterprise cert-manager for generating the user-specified certificate(secret) without Citadel or Istiod CA.
In istio-1.5.4, the user-specified certificate can be mounted via istio-certs for every Istio component with Citadel disabled. The mechanism was available to either Helm or operator installation.
- name: istio-certs
secret:
secretName: used-specified.istio-pilot-service-account
optional: true

      volumeMounts:
      - name: istio-certs
        mountPath: /etc/certs
        readOnly: true

In istio-1.7.4, it only allows user-specified certificates for ingressgateway/egressgateway with mountMtleCerts: true, but not istiod(pilot). Both ingressgateway/egressgateway show the following error when user-specified certificate is used in proxy:
2020-11-16T21:27:13.372694Z error xdsproxy failed to connect to upstream istiod.istio-system.svc:15012: context deadline exceeded
2020-11-16T14:27:13.373211Z warning envoy config StreamAggregatedResources gRPC config stream closed: 2, context deadline exceeded
2020-11-16T21:27:14.592114Z warn Envoy proxy is NOT ready: config not received from Pilot (is Pilot running?): cds updates: 0 successful, 0 rejected; lds updates: 0 successful, 0 rejected

The documentation is not clear to me how/when to use the pilotCertProvider, mountMtlsCerts, and dns certificate along with user-specified certificate. I cannot use plugging in existing CA certificates with Istiod CA.

Sorry if this reported issue is not a bug. Any help will be appreciated…

thanks,
Patrick

[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
client version: 1.7.4
control plane version: 1.9-alpha.7c7cd01f33072700961caebc0610e39ac2ac7aae
data plane version: none

**How was Istio installed?**
istioOperator 

**Environment where bug was observed (cloud vendor, OS, etc)**
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.9", GitCommit:"4fb7ed12476d57b8437ada90b4f93b17ffaeed99", GitTreeState:"clean", BuildDate:"2020-07-15T16:18:16Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.9", GitCommit:"4fb7ed12476d57b8437ada90b4f93b17ffaeed99", GitTreeState:"clean", BuildDate:"2020-07-15T16:10:45Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

**Comment by liwenhao0810** [comment]
@shankgan Could you please take a look at this issue.

**Comment by howardjohn** [comment]
I am a bit confused - the issue text says you installed 1.7.4 but the logs/istioctl output shows 1.9-alpha.7c7cd01f33072700961caebc0610e39ac2ac7aae. Are you intentionally running unreleased builds from `master` branch?

**Comment by patrickppeng** [comment]
No. I downloaded the istio-1.7.4 with "curl -L https://istio.io/downloadIstio | sh -" and the images from gcr.io/istio-testing:pilot:latest and gcr.io/istio-testing:proxyv2:latest. Am I using the incorrect image version?

**Comment by howardjohn** [comment]
How did you install after downloading?

**Comment by patrickppeng** [comment]
I tagged the images and put them to our own docker repository and using istio-operator to install the istio components. Attached has the istiooperator.yaml.
[operator_yaml.txt](https://github.com/istio/istio/files/5551049/operator_yaml.txt)

**Comment by howardjohn** [comment]
You should use istio/pilot:1.7.4, not gcr.io/istio-testing/pilot:latest.
The latter are unstable test builds.

**Comment by patrickppeng** [comment]
Thanks. I'll pull the correct pilot and proxyv2 images and test again.

**Comment by patrickppeng** [comment]
Pulled the following images and still seeing the "1.9-alpha" version...
    docker pull istio/pilot:1.7.4
    docker pull istio/proxyv2:1.7.4
    docker pull istio/operator:1.7.4
[root@flex188-33 istio-1.7.4]# istioctl version
client version: 1.7.4
control plane version: 1.9-alpha.7c7cd01f33072700961caebc0610e39ac2ac7aae
data plane version: 1.9-alpha.7c7cd01f33072700961caebc0610e39ac2ac7aae (4 proxies)

I'm still experiencing the reported issue with the new images. In the old istio version (1.5.4), it allowed to disable Citadel and apply user-specified secret to all istio components with mounted volume "istio-certs". In the latest Istio, user-specified secret can be added to ingress/egress proxies in the same way, but not available for pilot/istiod.  It ends up that proxies failed to connect to istiod:
2020-11-17T17:17:35.046783Z	error	xdsproxy	failed to connect to upstream istiod.istio-system.svc:15012: context deadline exceeded
Could you help to point out how to use user-specified certificate/secret without using istiod CA in the latest build?

**Comment by myidpt** [comment]
Shankar and I can help take a look in V1.7. But I'm very confused by the control plane version in your cluster.

**Comment by patrickppeng** [comment]
The old version of pilot in 1.5.4 has an istio-proxy that takes the mounted volume "istio-certs" for user-specified secret/certificate. The new istiod has a single discovery container and the documentation is not clear how to assign it an user-specified secret.

**Comment by shankgan** [comment]
Hi Patrick. 

"In the old istio version (1.5.4), it allowed to disable Citadel and apply user-specified secret to all istio components with mounted volume "istio-certs". In the latest Istio, user-specified secret can be added to ingress/egress proxies in the same way, but not available for pilot/istiod" - From the yaml file you provided, you appear to use both cacerts as well as istio-certs. Are they signed with the same signing key? More specifically, are they signed with the same **root-ca**? Istiod uses a cert created using the cacert signing key for its gRPC server.

To help debug why the output of `istioctl version` is showing you _"control plane version: 1.9-alpha.7c7cd01f33072700961caebc0610e39ac2ac7aae"_ , could you just give us a kubectl description (kubectl describe) of your (a) istiod deployment (b) your gateway deployment and (c) the deployment of one of your pods. 

Would also help if you could give us further details regarding the exact istioctl/helm commands you used to install istiod.


        


 For using a user-specified certificate/secret, can you just mount the "cacert" option?

**Comment by patrickppeng** [comment]
Hi Shankar,
I only tried the "istio-certs" for istiod with no luck and I didn't use cacerts. The reason you saw it in the istiooperator CR yaml because I have to overlay for additional volume/volumeMounts. Using cacerts is likely not acceptable by our security policy because customer may not be comfortable to provide root-cert and key to a 3rd party cert manager, similar reason to self-signing certs provided by 3rd party.

Attached has the description of deployment and pod
[ingress-pod.txt](https://github.com/istio/istio/files/5557108/ingress-pod.txt)
[istiod-deployment.txt](https://github.com/istio/istio/files/5557109/istiod-deployment.txt)
[gateway-deployment.txt](https://github.com/istio/istio/files/5557110/gateway-deployment.txt)
It seems like the operator inserts some labels during istiooperator installation that includes the "1.9.alpha" version.

Basically I created the istio-operator chart that includes the istiooperator CR under /templates and use helm to install it 
helm install istio-operator /opt/avaya/flex/clusters/cluster-env2-patpeng/workdir/flex-install-wwdwmbtd --namespace istio-system

**Comment by patrickppeng** [comment]
Hi Shankar,
Is it possible to totally bypass istiod CA and use used-specified secret/certificate for istiod/pilot and proxies in 1.7.4? or we must use cacerts for istiod/pilot?
What if we put istiod/pilot in a different namespace and inject a sidecar? Could that be a workaround to use "istio-certs" same as the old way? Any concerns or suggestions?

**Comment by patrickppeng** [comment]
I successfully installed istiod and ingressgateway/egressgateway on different namespace istio-control and istio-system with istiooperator CR. But the proxies failed to connect to discovery with the same error:
2020-11-20T00:18:12.527890Z     error   xdsproxy        failed to connect to upstream istiod.istio-system.svc:15012: context deadline exceeded
It seems like the CA is configured correctly to istiod.istio-control.svc:15012 with the "global.istioNamespace" set to "istio-control". But the discoveryAddress is still pointing to "discoveryAddress: istiod.istio-system.svc:15012" with wrong namespace. Did I miss any configuration?

here is the ingressgateway logs:
2020-11-20T00:17:48.349160Z     info    FLAG: --concurrency="0"
2020-11-20T00:17:48.349223Z     info    FLAG: --domain="istio-system.svc.cluster.local"
2020-11-20T00:17:48.349231Z     info    FLAG: --help="false"
2020-11-20T00:17:48.349245Z     info    FLAG: --log_as_json="false"
2020-11-20T00:17:48.349253Z     info    FLAG: --log_caller=""
2020-11-20T00:17:48.349260Z     info    FLAG: --log_output_level="default:info"
2020-11-20T00:17:48.349269Z     info    FLAG: --log_rotate=""
2020-11-20T00:17:48.349275Z     info    FLAG: --log_rotate_max_age="30"
2020-11-20T00:17:48.349283Z     info    FLAG: --log_rotate_max_backups="1000"
2020-11-20T00:17:48.349295Z     info    FLAG: --log_rotate_max_size="104857600"
2020-11-20T00:17:48.349301Z     info    FLAG: --log_stacktrace_level="default:none"
2020-11-20T00:17:48.349310Z     info    FLAG: --log_target="[stdout]"
2020-11-20T00:17:48.349321Z     info    FLAG: --meshConfig="./etc/istio/config/mesh"
2020-11-20T00:17:48.349326Z     info    FLAG: --outlierLogPath=""
2020-11-20T00:17:48.349334Z     info    FLAG: --proxyComponentLogLevel="misc:error"
2020-11-20T00:17:48.349346Z     info    FLAG: --proxyLogLevel="warning"
2020-11-20T00:17:48.349352Z     info    FLAG: --serviceCluster="istio-ingressgateway"
2020-11-20T00:17:48.349357Z     info    FLAG: --stsPort="0"
2020-11-20T00:17:48.349361Z     info    FLAG: --templateFile=""
2020-11-20T00:17:48.349369Z     info    FLAG: --tokenManagerPlugin="GoogleTokenExchange"
2020-11-20T00:17:48.349378Z     info    Version 1.9-alpha.7c7cd01f33072700961caebc0610e39ac2ac7aae-7c7cd01f33072700961caebc0610e39ac2ac7aae-Clean
2020-11-20T00:17:48.349572Z     info    Obtained private IP [10.200.36.140]
2020-11-20T00:17:48.350680Z     info    Effective config: binaryPath: /usr/local/bin/envoy
concurrency: 0
configPath: ./etc/istio/proxy
controlPlaneAuthPolicy: MUTUAL_TLS
discoveryAddress: istiod.istio-system.svc:15012
drainDuration: 45s
envoyAccessLogService: {}
envoyMetricsService: {}
parentShutdownDuration: 60s
proxyAdminPort: 15000
serviceCluster: istio-ingressgateway
statNameLength: 189
statusPort: 15020
terminationDrainDuration: 5s
tracing:
  zipkin:
    address: zipkin.istio-system:9411

2020-11-20T00:17:48.350755Z     info    Proxy role: &model.Proxy{RWMutex:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, Type:"router", IPAddresses:[]string{"10.200.36.140"}, ID:"istio-ingressgateway-869c844d65-vtsv5.istio-system", Locality:(*envoy_config_core_v3.Locality)(nil), DNSDomain:"istio-system.svc.cluster.local", ConfigNamespace:"", Metadata:(*model.NodeMetadata)(nil), SidecarScope:(*model.SidecarScope)(nil), PrevSidecarScope:(*model.SidecarScope)(nil), MergedGateway:(*model.MergedGateway)(nil), ServiceInstances:[]*model.ServiceInstance(nil), IstioVersion:(*model.IstioVersion)(nil), VerifiedIdentity:(*spiffe.Identity)(nil), ipv6Support:false, ipv4Support:false, GlobalUnicastIP:"", XdsResourceGenerator:model.XdsResourceGenerator(nil), WatchedResources:map[string]*model.WatchedResource(nil)}
2020-11-20T00:17:48.350763Z     info    JWT policy is first-party-jwt
2020-11-20T00:17:48.350836Z     info    PilotSAN []string{"istiod.istio-system.svc"}
2020-11-20T00:17:48.350892Z     info    sa.serverOptions.CAEndpoint == istiod.istio-control.svc:15012 Citadel
2020-11-20T00:17:48.350954Z     info    Using CA istiod.istio-control.svc:15012 cert with certs: var/run/secrets/istio/root-cert.pem

**Comment by shankgan** [comment]
Hi Patrick, 

> looking for help on how to use user-specified certificate for in-cluster mtls communication.

Which communication are you referring to? 
Istiod <-> Istio Proxy (Control-Plane)?
Istiod <-> Ingress/Egress Gateway (Control Plane),
Istio proxy<-> Ingress/Egress gateway(Data-Plane),
Istio-proxy <-> Istio proxy (Data plane) 
All of the above?

> Is it possible to totally bypass istiod CA and use used-specified secret/certificate for istiod/pilot and proxies in 1.7.4? or we must use cacerts for istiod/pilot?

User specified certificate can be used for istiod + ingress + egress gateway. However, this is much more complex for proxies and not recommended to do *manually*. For instance, the istio-proxy certificate identity is tied to the kubernetes service account. Are you planning to create signed certificate (from the third party CA) for evey pod in every single service account? I don't think there is a straightforward mechanism for Istiod to inject this custom created certificate in every istio-proxy anyway.

> 2020-11-20T00:17:48.350954Z info Using CA istiod.istio-control.svc:15012 cert with certs: var/run/secrets/istio/root-cert.pem

Looks like the configured certificates are not being picked up. Can you change your global.pilotcertprovider to "custom". That should atleast ensure that the connectivity between the control plane components should go through.

**Comment by shankgan** [comment]
@patrickppeng - let us know if you are still having any issues with connection. Else, we can close the Issue

**Comment by patrickppeng** [comment]
Sorry for the delay... just back from US holiday.
Yes, we are still experiencing the same issue with global.mountMtlsCerts to "true" and user-specified certificates for all the following paths
Istiod <-> Istio Proxy (Control-Plane)
Istiod <-> Ingress/Egress Gateway (Control Plane),
Istio proxy<-> Ingress/Egress gateway(Data-Plane),
Istio-proxy <-> Istio proxy (Data plane)
external client <-> Ingress/Egress Gateway

Yes, our cert manager CA is able to create certificate for istio-proxy with different kubernetes service account.

I will try your suggestion to set global.pilotcertprovider to "custom".

**Comment by patrickppeng** [comment]
Istiod failed to come up after set global.pilotcertprovider to "custom" and mounted user-specified certificate (cert-chain.pem) to /var/run/secrets/istio-dns.

2020-11-30T22:09:03.958207Z	info	rootcertrotator	Set up back off time 39m0s to start rotator.
2020-11-30T22:09:03.958246Z	info	initializing Istiod DNS certificates host: istiod.istio-system.svc, custom host: 
2020-11-30T22:09:03.958274Z	info	User specified cert provider: custom
2020-11-30T22:09:03.958692Z	info	rootcertrotator	Jitter is enabled, wait 39m0s before starting root cert rotator.
2020-11-30T22:09:03.958772Z	info	adding watcher for certificate var/run/secrets/istio-dns/cert-chain.pem
2020-11-30T22:09:03.958949Z	info	adding watcher for certificate var/run/secrets/istio-dns/key.pem
2020-11-30T22:09:03.959159Z	info	spiffe	Added 1 certs to trust domain cluster.local in peer cert verifier
2020-11-30T22:09:03.959210Z	info	initializing secure discovery service
2020-11-30T22:09:03.959491Z	info	using max conn age of 30m0s
2020-11-30T22:09:03.959544Z	info	initializing secure webhook server for istiod webhooks
2020-11-30T22:09:03.959625Z	info	initializing sidecar injector
2020-11-30T22:09:04.047396Z	info	initializing config validator
Error: failed to create discovery service: error initializing config validator: no such file or directory
2020-11-30T22:09:04.047773Z	error	failed to start validation controller: no such file or directory
2020-11-30T22:09:04.047825Z	error	failed to create discovery service: error initializing config validator: no such file or directory

**Comment by ratnadeep26** [comment]
Hi Shankar,

I am working with Patrick(patrickppeng) on this one.
Can you please clearly document if someone wants to use external certificates for 
Ingress Gateway
Egress Gateway
Istiod
Istio proxy
what are the helm variables that one would have to override and set and what are the secrets name(and the keys in those secrets) to place the external certificates.

I tried setting 
mountMtlsCerts: true
pilotCertProvider: custom
and creating secrets with external certs as per secrets names defined in operator.yaml. but I could not get istiod to come up. It still looks for certificate at location var/run/secrets/istio-dns/cert-chain.pem

error   error initializing certificate watches: open var/run/secrets/istio-dns/cert-chain.pem: no such file or directory

**Comment by patrickppeng** [comment]
Hi Shankar,
we know proxy (ingress, egress, & sidecar) can take external certs by setting "mountMtlsCerts: true" but it is still confusing how istiod takes external cert with "pilotCertProvider: custom". what certs need to be included in the secret and what path needs to be mounted to?

**Comment by patrickppeng** [comment]
Ok, after set pilotCertProvider to "custom", we have to add the following args to discovery container to support the user specified certificates that are mounted to /etc/certs:
  spec:
      containers:
      - args:
        - discovery
        - --monitoringAddr=:15014
        **- --tlsKeyFile=/etc/certs/key.pem
        - --tlsCertFile=/etc/certs/cert-chain.pem
        - --caCertFile=/etc/certs/root-cert.pem**
